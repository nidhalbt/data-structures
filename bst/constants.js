export const NIL_NODE = null;
export const NIL_VALUE = null;

export const isNil = (node) => {
  return !node;
};
