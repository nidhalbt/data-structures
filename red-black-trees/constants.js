export const COLORS = {
  RED: 'red',
  BLACK: 'black'
};
