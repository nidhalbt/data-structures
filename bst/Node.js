import { isNil, NIL_NODE } from "./constants";

export class Node {
  constructor(value, parent, left, right) {
    this.value = value;
    this.parent = parent;
    this.left = left;
    this.right = right;
    this.key = Math.random() * (1 - Math.random()) * Date.now();
  }

  newChild(value) {
    return new Node(value, this, NIL_NODE, NIL_NODE);
  }

  equals(node) {
    return !isNil(node) && (this.key === node.key);
  }

  push(value) {
    if (!this.value) {
      this.value = value;
      return this;
    }

    if (value < this.value) {
      return this.pushLeft(value);
    } else {
      return this.pushRight(value);
    }
  }

  pushLeft(value) {
    if (isNil(this.left)) {
      this.left = this.newChild(value);
      return this.left;
    }

    return this.left.push(value);
  }

  pushRight(value) {
    if (isNil(this.right)) {
      this.right = this.newChild(value);
      return this.right;
    }

    return this.right.push(value);
  }

  reverseLineage(child) {
    if (isNil(child)) {
      throw new Error('reverseLineage should NOT be called with an empty node');
    }

    if (isNil(child.parent) || !child.parent.equals(this)) {
      throw new Error('reverseLineage should not called on parent node to switch with its immediate child');
    }

    if (this.isLeftChild) {
      this.parent.left = child;
    }
    if (this.isRightChild) {
      this.parent.right = child;
    }

    child.parent = this.parent;
    this.parent = child;
  }

  rotateRight() {
    const { left } = this;
    this.reverseLineage(left);

    this.left = left.right;
    if (!isNil(this.left)) {
      this.left.parent = this;
    }
    left.right = this;
  }

  rotateLeft() {
    const { right } = this;
    this.reverseLineage(right);

    this.right = right.left;
    if (!isNil(this.right)) {
      this.right.parent = this;
    }
    right.left = this;
  }

  stripKeys() {
    delete this.key;
    if (!isNil(this.left)) {
      this.left.stripKeys();
    }
    if (!isNil(this.right)) {
      this.right.stripKeys();
    }
  }

  get isRoot() {
    return isNil(this.parent);
  }

  get isLeftChild() {
    return !isNil(this.parent) && this.equals(this.parent.left);
  }

  get isRightChild() {
    return !isNil(this.parent) && this.equals(this.parent.right);
  }

  get uncle() {
    const { isRoot, parent, grandparent } = this;
    if (isRoot || parent.isRoot) {
      return NIL_NODE;
    }

    return parent.isLeftChild ? grandparent.right : grandparent.left;
  }

  get grandparent() {
    if (isNil(this.parent) || isNil(this.parent.parent)) {
      return NIL_NODE;
    }

    return this.parent.parent;
  }
}
