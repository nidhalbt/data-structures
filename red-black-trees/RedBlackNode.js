import { isNil, NIL_NODE } from "../bst/constants";
import { Node } from "../bst/Node";
import { COLORS } from "./constants";

export class RedBlackNode extends Node {
  constructor(value, parent, left, right, color) {
    super(value, parent, left, right);
    this.color = color;
  }

  static reColor(node) {
    if (node.isRoot) {
      node.color = COLORS.BLACK;
      return node;
    }

    const { uncle, parent, grandparent, uncleIsBlack, uncleIsRed } = node;

    if (parent.isRed) {
      if (uncleIsRed) {
        parent.color = COLORS.BLACK;
        uncle.color = COLORS.BLACK;
        // if uncle is not Nil a node than grandparent is guaranteed to not be Nil
        grandparent.color = COLORS.RED;
        return RedBlackNode.reColor(grandparent);
      } else if (uncleIsBlack) {
        if (parent.isLeftChild && node.isLeftChild) {
          grandparent.rotateRight();
          grandparent.exchangeColors(parent);
          return grandparent.getRootNode();
        }
        if (parent.isLeftChild && node.isRightChild) {
          parent.rotateLeft();
          grandparent.rotateRight();
          grandparent.exchangeColors(node);
          return grandparent.getRootNode();
        }
        if (parent.isRightChild && node.isRightChild) {
          grandparent.rotateLeft();
          grandparent.exchangeColors(parent);
          return grandparent.getRootNode();
        }
        if (parent.isRightChild && node.isLeftChild) {
          parent.rotateRight();
          grandparent.rotateLeft();
          grandparent.exchangeColors(node);
          return grandparent.getRootNode();
        }
      }
    }

    return node.getRootNode();
  }

  newChild(value) {
    return new RedBlackNode(value, this, NIL_NODE, NIL_NODE, COLORS.RED);
  }

  exchangeColors(node) {
    const color = this.color;
    this.color = node.color;
    node.color = color;
  }

  getRootNode() {
    if (this.isRoot) {
      return this;
    }
    return this.parent.getRootNode();
  }

  get isRed() {
    return this.color === COLORS.RED;
  }

  get isBlack() {
    return this.color === COLORS.BLACK;
  }

  get uncleIsBlack() {
    const { uncle, grandparent } = this;
    // nodes which are not initialized are considered black
    return (grandparent && isNil(uncle)) || (!isNil(uncle) && uncle.isBlack);
  }

  get uncleIsRed() {
    const { uncle } = this;
    return !isNil(uncle) && uncle.isRed;
  }
}
