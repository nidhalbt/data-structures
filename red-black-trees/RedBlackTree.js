import { NIL_NODE, NIL_VALUE } from "../bst/constants";
import { COLORS } from "./constants";
import { RedBlackNode } from "./RedBlackNode";

export class RedBlackTree {
  constructor() {
    this.root = new RedBlackNode(NIL_VALUE, NIL_NODE, NIL_NODE, NIL_NODE, COLORS.BLACK);
  }

  batch(values) {
    this.root = new RedBlackNode(NIL_VALUE, NIL_NODE, NIL_NODE, NIL_NODE);

    values.forEach((value) => {
      const node = this.root.push(value);
      this.root = RedBlackNode.reColor(node);
    });
  }

  one(value) {
    const node = this.root.push(value);
    this.root = this.root.getRootNode();
    return node;
  }
}
