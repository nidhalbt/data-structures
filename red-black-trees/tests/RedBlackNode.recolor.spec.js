import { NIL_NODE, NIL_VALUE } from "../../bst/constants";
import { COLORS } from "../constants";
import { RedBlackNode } from "../RedBlackNode";
import { RedBlackTree } from "../RedBlackTree";

describe('RedBlackNode.reColor', () => {
  let root, left, right, leftLeft, leftRight, rightLeft, rightRight;

  describe('root', () => {
    it('should set root color to black', () => {
      const node = new RedBlackNode();
      node.push(24);
      RedBlackNode.reColor(node);
      expect(node.isBlack).toBeTruthy();
    });
  });

  describe('red uncle', () => {
    const bfe = () => {
      root = new RedBlackNode(10, NIL_NODE, NIL_NODE, NIL_NODE, COLORS.BLACK);

      left = new RedBlackNode(10, root, NIL_NODE, NIL_NODE, COLORS.RED);
      root.left = left;

      right = new RedBlackNode(10, root, NIL_NODE, NIL_NODE, COLORS.RED);
      root.right = right;
    };

    beforeEach(bfe);

    it('should change parent and uncle color to black, grandparent to red, then retry from grandparent', () => {
      const node = root.push(5);
      RedBlackNode.reColor(node);
      expect(node.isRed).toBeTruthy();
      expect(root.isBlack).toBeTruthy();
      expect(left.isBlack).toBeTruthy();
      expect(right.isBlack).toBeTruthy();
      expect(left.left.isRed).toBeTruthy();
    });
  });

  describe('black uncle', () => {
    it('left left case', () => {
      const ten = new RedBlackNode(NIL_VALUE, NIL_NODE, NIL_NODE, NIL_NODE);
      ten.push(10);
      RedBlackNode.reColor(ten);

      const nine = ten.push(9);
      RedBlackNode.reColor(nine);

      const eight = ten.push(8);
      RedBlackNode.reColor(eight);

      expect(nine.isRoot).toBeTruthy();
      expect(nine.right).toEqual(ten);
      expect(nine.left).toEqual(eight);
      expect(nine.isBlack).toBeTruthy();
      expect(ten.isRed).toBeTruthy();
      expect(eight.isRed).toBeTruthy();
    });

    it('left right case', () => {
      const ten = new RedBlackNode(NIL_VALUE, NIL_NODE, NIL_NODE, NIL_NODE);
      ten.push(10);
      RedBlackNode.reColor(ten);

      const eight = ten.push(8);
      RedBlackNode.reColor(eight);

      const nine = ten.push(9);
      RedBlackNode.reColor(nine);

      expect(nine.isRoot).toBeTruthy();
      expect(nine.right).toEqual(ten);
      expect(nine.left).toEqual(eight);
      expect(nine.isBlack).toBeTruthy();
      expect(ten.isRed).toBeTruthy();
      expect(eight.isRed).toBeTruthy();
    });

    it('right left case', () => {
      const eight = new RedBlackNode(NIL_VALUE, NIL_NODE, NIL_NODE, NIL_NODE);
      eight.push(8);
      RedBlackNode.reColor(eight);
      const ten = eight.push(10);
      RedBlackNode.reColor(ten);
      const nine = eight.push(9);
      RedBlackNode.reColor(nine);

      expect(nine.isRoot).toBeTruthy();
      expect(nine.right).toEqual(ten);
      expect(nine.left).toEqual(eight);
      expect(nine.isBlack).toBeTruthy();
      expect(ten.isRed).toBeTruthy();
      expect(eight.isRed).toBeTruthy();
    });

    it('right right case', () => {
      const ten = new RedBlackNode(NIL_VALUE, NIL_NODE, NIL_NODE, NIL_NODE);
      ten.push(10);
      RedBlackNode.reColor(ten);

      const eleven = ten.push(11);
      RedBlackNode.reColor(eleven);

      const twelve = ten.push(12);
      RedBlackNode.reColor(twelve);

      expect(eleven.isRoot).toBeTruthy();
      expect(eleven.left).toEqual(ten);
      expect(eleven.right).toEqual(twelve);
      expect(eleven.isBlack).toBeTruthy();
      expect(ten.isRed).toBeTruthy();
      expect(twelve.isRed).toBeTruthy();
    });
  });

  describe('different benchmarks', () => {
    it('should match a known benchmark', () => {
      const values = [0, 12, 9, 0, 1, 35, 31, 3, 43, 37, 7, 36, 17, 2, 18, 32, 26, 12, 34, 9, 44, 37, 47, 9, 21, 37, 38, 45, 44, 18];
      const tree = new RedBlackTree();
      tree.batch(values);

      expect(tree.root.color).toEqual(COLORS.BLACK);
      expect(tree.root.value).toEqual(31);

      expect(tree.root.left.color).toEqual(COLORS.RED);
      expect(tree.root.left.value).toEqual(9);

      expect(tree.root.left.right.color).toEqual(COLORS.BLACK);
      expect(tree.root.left.right.value).toEqual(17);

      expect(tree.root.left.right.right.color).toEqual(COLORS.RED);
      expect(tree.root.left.right.right.value).toEqual(21);

      expect(tree.root.right.left.left.right.color).toEqual(COLORS.RED);
      expect(tree.root.right.left.left.right.value).toEqual(34);

      expect(tree.root.right.color).toEqual(COLORS.RED);
      expect(tree.root.right.value).toEqual(37);

      expect(tree.root.right.right.left.left.color).toEqual(COLORS.RED);
      expect(tree.root.right.right.left.left.value).toEqual(37);

      expect(tree.root.right.left.right.color).toEqual(COLORS.BLACK);
      expect(tree.root.right.left.right.value).toEqual(36);
    });

    it('should match snapshot', () => {
      const values = [43, 34, 19, 37, 36, 24, 12, 41, 30, 40, 35, 50, 27, 48, 35, 13, 43, 3, 7, 46, 31, 27, 5, 43, 46, 27, 47, 45, 3, 29];
      const tree = new RedBlackTree();
      tree.batch(values);
      tree.root.stripKeys();
      expect(tree.root).toMatchSnapshot();
    });
  });
});
