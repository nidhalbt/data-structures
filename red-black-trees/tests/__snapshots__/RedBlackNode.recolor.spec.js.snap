// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RedBlackNode.reColor different benchmarks should match snapshot 1`] = `
RedBlackNode {
  "color": "black",
  "left": RedBlackNode {
    "color": "red",
    "left": RedBlackNode {
      "color": "black",
      "left": RedBlackNode {
        "color": "red",
        "left": RedBlackNode {
          "color": "black",
          "left": null,
          "parent": [Circular],
          "right": RedBlackNode {
            "color": "red",
            "left": null,
            "parent": [Circular],
            "right": null,
            "value": 3,
          },
          "value": 3,
        },
        "parent": [Circular],
        "right": RedBlackNode {
          "color": "black",
          "left": null,
          "parent": [Circular],
          "right": null,
          "value": 7,
        },
        "value": 5,
      },
      "parent": [Circular],
      "right": RedBlackNode {
        "color": "black",
        "left": null,
        "parent": [Circular],
        "right": null,
        "value": 13,
      },
      "value": 12,
    },
    "parent": [Circular],
    "right": RedBlackNode {
      "color": "black",
      "left": RedBlackNode {
        "color": "black",
        "left": null,
        "parent": [Circular],
        "right": null,
        "value": 24,
      },
      "parent": [Circular],
      "right": RedBlackNode {
        "color": "red",
        "left": RedBlackNode {
          "color": "black",
          "left": RedBlackNode {
            "color": "red",
            "left": null,
            "parent": [Circular],
            "right": null,
            "value": 27,
          },
          "parent": [Circular],
          "right": RedBlackNode {
            "color": "red",
            "left": null,
            "parent": [Circular],
            "right": null,
            "value": 29,
          },
          "value": 27,
        },
        "parent": [Circular],
        "right": RedBlackNode {
          "color": "black",
          "left": null,
          "parent": [Circular],
          "right": null,
          "value": 31,
        },
        "value": 30,
      },
      "value": 27,
    },
    "value": 19,
  },
  "parent": null,
  "right": RedBlackNode {
    "color": "red",
    "left": RedBlackNode {
      "color": "black",
      "left": RedBlackNode {
        "color": "black",
        "left": RedBlackNode {
          "color": "red",
          "left": null,
          "parent": [Circular],
          "right": null,
          "value": 35,
        },
        "parent": [Circular],
        "right": RedBlackNode {
          "color": "red",
          "left": null,
          "parent": [Circular],
          "right": null,
          "value": 36,
        },
        "value": 35,
      },
      "parent": [Circular],
      "right": RedBlackNode {
        "color": "black",
        "left": null,
        "parent": [Circular],
        "right": null,
        "value": 40,
      },
      "value": 37,
    },
    "parent": [Circular],
    "right": RedBlackNode {
      "color": "black",
      "left": RedBlackNode {
        "color": "red",
        "left": RedBlackNode {
          "color": "black",
          "left": null,
          "parent": [Circular],
          "right": null,
          "value": 43,
        },
        "parent": [Circular],
        "right": RedBlackNode {
          "color": "black",
          "left": null,
          "parent": [Circular],
          "right": RedBlackNode {
            "color": "red",
            "left": null,
            "parent": [Circular],
            "right": null,
            "value": 45,
          },
          "value": 43,
        },
        "value": 43,
      },
      "parent": [Circular],
      "right": RedBlackNode {
        "color": "red",
        "left": RedBlackNode {
          "color": "black",
          "left": null,
          "parent": [Circular],
          "right": RedBlackNode {
            "color": "red",
            "left": null,
            "parent": [Circular],
            "right": null,
            "value": 47,
          },
          "value": 46,
        },
        "parent": [Circular],
        "right": RedBlackNode {
          "color": "black",
          "left": null,
          "parent": [Circular],
          "right": null,
          "value": 50,
        },
        "value": 48,
      },
      "value": 46,
    },
    "value": 41,
  },
  "value": 34,
}
`;
